### PR TITLE
[ZEPPELIN-3001] Disable ace editor code completion when backend interpreter support completion

### PR DESCRIPTION
**Pull request by prabhjyotsingh**
### What is this PR for?
This is what I see in zeppelin and jupyter. The difference is due to zeppelin also add code completion of ace editor. I think code completion of ace editor should only be enabled when the backend interpreter doesn't support completion.

### What type of PR is it?
[Bug Fix]

### What is the Jira issue?
* [ZEPPELIN-3001](https://issues.apache.org/jira/browse/ZEPPELIN-3001)

### Questions:
* Does the licenses files need update? N/A
* Is there breaking changes for older versions? N/A
* Does this needs documentation? N/A

### CODE REVIEW

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 1]
```diff
@@ -1393,7 +1393,13 @@ private void completion(NotebookSocket conn, HashSet<String> userAndRoles, Noteb
     }
 
     final Note note = notebook.getNote(getOpenNoteId(conn));
-    List<InterpreterCompletion> candidates = note.completion(paragraphId, buffer, cursor);
```
Comment (prabhjyotsingh) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: this used to throw exception, I think its a better idea to log that error and send empty array to frontend,

File: spark/interpreter/src/main/resources/interpreter-setting.json [hunk 2]
```diff
@@ -227,7 +232,8 @@
     },
     "editor": {
       "language": "r",
-      "editOnDblClick": false
+      "editOnDblClick": false,
+      "completionSupport": true
```
Comment (zjffdu) [inline spark/interpreter/src/main/resources/interpreter-setting.json]: Should be false for SparkRInterpreter, as the completion method of SparkRInterpreter always return empty list

File: zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js [hunk 3]
```diff
@@ -963,8 +963,73 @@ function ParagraphCtrl($scope, $rootScope, $route, $window, $routeParams, $locat
     }
   };
 
+  // ref: https://github.com/ajaxorg/ace/blob/5021d0193d9f2bba5a978d0b1d7a4f73d18ce713/lib/ace/autocomplete.js#L454
+  const completionSupportWithoutBackend = function(str) {
```
Comment (prabhjyotsingh) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js]: This function is copied from the above link, and the only difference between this and below function is 
```
if (item.meta !== '') {
  return false;
}
```

File: spark/interpreter/src/main/resources/interpreter-setting.json (hunk 2 above)
Comment (prabhjyotsingh) [reply spark/interpreter/src/main/resources/interpreter-setting.json]: Sure, totally missed it, let me make that change.

**Comment by zjffdu** [comment]
@prabhjyotsingh There's no response when I type tab, anything broken ?

**Comment by prabhjyotsingh** [comment]
That sounds like nothing came back from the backend, can you check your WebSocket messages for "op= COMPLETION_LIST", then look under data.completions, that should be an empty array. 

Right now this PR is suppressing the default (BasicAutocompletion) completion if `completionSupport` is set to `true` for a given interpreter, should I change this to, if the return list is empty, then show the basic autocompletion?

**Comment by prabhjyotsingh** [comment]
@felixcheung agreed, but at least it gives back something basic.

**Comment by zjffdu** [comment]
@prabhjyotsingh What I see is that the paragraphId sent from frontend is incorrect which cause the issue I mention in the last comment. When I type tab in the first paragraph, it would send the last pargraphId to backend. could you help check that ?

**Comment by zjffdu** [comment]
@prabhjyotsingh I try the latest PR, but it doesn't work when I enable zeppelin.spark.useNew. I can confirm the completion api return the same thing no matter whether zeppelin.spark.useNew is true or false. Could you help check what's difference in frontend ? Thanks

**Comment by zjffdu** [comment]
ping @prabhjyotsingh

**Comment by prabhjyotsingh** [comment]
I didn't know this, there is a difference is spark, let me try it out.

**Comment by prabhjyotsingh** [comment]
@zjffdu I was able to reproduce the issue that you mentioned, have made a fix for the same.

**Comment by zjffdu** [comment]
Thanks @prabhjyotsingh , but it still doesn't work for me.

**Comment by zjffdu** [comment]
Never mind, it works for me now. Thanks @prabhjyotsingh  +1

**Comment by prabhjyotsingh** [comment]
Thanks for the review will merge this if no more discussion.

**Pull request closed without merging**